### ISSUE TITLE
Add Sapling support to wallet RPCs

### ISSUE THREAD

**Issue by str4d** [zcash/zcash#3063]
- [ ] `z_listreceivedbyaddress`
- [ ] `z_getbalance`
- [ ] `z_gettotalbalance`
- [ ] `z_sendmany`
- [ ] `z_shieldcoinbase`
- [ ] `z_mergetoaddress`

**Comment by daira** [comment]
We need to decide how and when diversified addresses are used by RPC calls:

* Should ``z_listaddresses`` return multiple diversified addresses for the same spend authority, or only one? If the latter, how should it pick?
* ``z_getnewaddress`` should clearly return an address with a fresh spend authority, because that is what happened in Sprout and the user might be depending on it. Is it necessary to have a different call that returns a diversified address for an existing spend authority, and if so how should it specify which authority?
* ``z_listreceivedbyaddress`` has an obvious semantics, i.e. filter by the diversified address given in the note plaintext, but this doesn't expose all of the efficiency advantages of diversified addresses. (The node only has to scan the block chain once, but you still need to make a separate call per diversified address, of which there may be many, to get all of the transactions.) Maybe there needs to be a ``z_listallreceived`` or ``z_listreceivedfromheight`` call.

I favour having a namespace of user-meaningful names for spend authorities similar to what people thought the old "account" feature did. There would need to be a call to list these spend authority identifiers and the known addresses associated with each of them.

**Comment by str4d** [comment]
> I also think we should have a `z_getdiversifiedaddress <existing_address>` call, which looks up the ivk for the given address and returns a new diversified address for that ivk.

This API should probably also just accept an IVK directly (or an IVK label if we implement those in future). I was also wondering whether it should accept a diversifier index, but I think not; given that we have diversifier derivation in ZIP 32, we should just maintain an internal counter per-IVK of the last diversified address generated, and increment it as necessary.

**Comment by daira** [comment]
> Do you mean by incrementing ``account`` [...] or by incrementing ``i`` [...]?

By incrementing ``i``.

**Comment by daira** [comment]
Now that I think about it some more, I'm torn. If we return addresses with a fresh spending authority, we're effectively forgoing the efficiency improvement of using diversified addresses by default. We *could* have ``z_getnewaddress`` return a new diversified address for the default spending authority of the account. I know I said above that Bitcoin-derived wallets might be depending on it being a fresh spending authority, but how likely is that actually?

**Comment by str4d** [comment]
If we want to do that, I propose the following change to the API:

`z_getnewaddress [type [account_num]]`

If `type == "sapling"`, this generates a new diversified address for
`account = account_num` (`default = 0`). If we run out of diversifiers
within that account number, we can either return an error, or start using
child spend authorities `i` under that account.

Alternatively we could use a text `label` instead of an account number. We
can't use `account` directly, as that's already used elsewhere in the RPC
(even though it's all disabled).

**Comment by daira** [comment]
We won't run out of diversifiers because at, say, 1 nanosecond per request, it would take 5 billion years to exhaust 2<sup>87</sup> diversifiers.

**Comment by str4d** [comment]
Oh, I don't know... We should probably make an RPC test that generates them
all, to make sure that it correctly returns an error at the end 😉

**Comment by daira** [comment]
I agree with this change, with ``type == "sapling"`` as the default.

**Comment by daira** [comment]
Sorry for spoiling your joke, but that's what mocks are for :-)

**Comment by bitcartel** [comment]
Sapling support for `z_importviewingkey` and `z_exportviewingkey` is not expected for v2.0.1 i.e. Sapling activation.

**Comment by bitcartel** [comment]
TXToJSON needs to be updated to return appropriate fields for Sapling data e.g. shielded spends, outputs, used by RPC calls like getrawtransaction to return contextual info.

**Comment by exarkun** [comment]
I see there is some discussion about how to expose diversified keys here.  Is this issue the best place for further such discussion or is there another issue for that functionality specifically?  (If not, would one be welcome?)
